File: tests/extra_libc.c
// Test libc EXTRA calls.
// FIXME: test file routines.
//
// (c) Reuben Thomas 1994-2018
//
// The package is distributed under the GNU Public License version 3, or,
// at your option, any later version.
//
// THIS PROGRAM IS PROVIDED AS IS, WITH NO WARRANTY. USE IS AT THE USER‘S
// RISK.

#include "tests.h"


#define BUFFER 0x100


int main(void)
{
    int exception = 0;

    // Data for ARGC/ARG_LEN/ARG_COPY tests
    int argc = 3;
    char *argv[] = {strdup("foo"), strdup("bard"), strdup("basilisk")};

    state * S = init_default_stacks(1024);
    assert(register_args(S, argc, argv) == 0);

    ass_number(S, OX_ARGC); ass_number(S, OXLIB_LIBC); ass_action(S, O_EXTRA);
    ass_number(S, 1); ass_number(S, OX_ARG_LEN); ass_number(S, OXLIB_LIBC); ass_action(S, O_EXTRA);
    ass_number(S, 1); ass_number(S, BUFFER); ass_number(S, 16);
    ass_number(S, OX_ARG_COPY); ass_number(S, OXLIB_LIBC); ass_action(S, O_EXTRA);

    // Test 1: OX_ARGC
    do {
        single_step(S);
    } while (S->I != O_EXTRA);
    WORD argc_;
    load_stack(S, 0, &argc_);
    printf("argc is %"PRI_WORD", and should be %d\n", argc_, argc);
    WORD temp;
    POP(&temp);
    if (temp != argc) {
       printf("Error in extra instructions tests: PC = %"PRI_UWORD"\n", S->PC);
        exit(1);
    }

    // Test 2: OX_ARG_LEN
    do {
        single_step(S);
    } while (S->I != O_EXTRA);
    UWORD arg1len;
    load_stack(S, 0, (WORD *)&arg1len);
    printf("arg 1's length is %"PRI_UWORD", and should be %zu\n", arg1len, strlen(argv[1]));
    if (arg1len != strlen(argv[1])) {
        printf("Error in extra instructions tests: PC = %"PRI_UWORD"\n", S->PC);
        exit(1);
    }

    // Tests 3 & 4: OX_ARG_COPY
    do {
        single_step(S);
    } while (S->I != O_EXTRA);
    POP((WORD *)&arg1len);
    printf("arg 1's length is %"PRI_UWORD", and should be %zu\n", arg1len, strlen(argv[1]));
    if (arg1len != strlen(argv[1])) {
        printf("Error in extra instructions tests: PC = %"PRI_UWORD"\n", S->PC);
        exit(1);
    }
    printf("arg 1 is %s, and should be %s\n", native_address_of_range(S, BUFFER, 0), argv[1]);
    if (strncmp((char *)native_address_of_range(S, BUFFER, 0), argv[1], strlen(argv[1])) != 0) {
        printf("Error in extra instructions tests: PC = %"PRI_UWORD"\n", S->PC);
        exit(1);
    }

    destroy(S);
    for (int i = 0; i < argc; i++)
        free(argv[i]);

    assert(exception == 0);
    printf("Extra instructions tests ran OK\n");
    return 0;
}

File: tests/extra_libc.py
# Test libc extra instructions.
# FIXME: test file routines.
#
# (c) Reuben Thomas 1994-2018
#
# The package is distributed under the MIT/X11 License.
#
# THIS PROGRAM IS PROVIDED AS IS, WITH NO WARRANTY. USE IS AT THE USER’S
# RISK.

from smite import *
VM = State()
VM.globalize(globals())


# Data for ARGC/ARGV tests
args = [b"foo", b"bard", b"basilisk"]
VM.register_args(*args)

# Test code
buffer = 0x100

# Put address of buffer on stack for later
number(buffer)
number(16) # arbitary number > strlen(args[1])
number(LIB_SMITE_CURRENT_STATE)
action(LIB_SMITE)
number(LIB_SMITE_NATIVE_ADDRESS_OF_RANGE)
action(LIB_SMITE)

# Test LIB_C_ARGC
number(LIB_C_ARGC)
action(LIB_C)

step(addr=VM.here + 1)
argc = S.pop()
print("argc is {}, and should be {}".format(argc, len(args)))
if argc != len(args):
    print("Error in extra instructions tests: PC = {:#x}".format(PC.get()))
    sys.exit(1)

# Test LIB_C_ARG
number(1)
number(LIB_C_ARG)
action(LIB_C)
for i in range(sizeof(c_char_p) // word_size):
    number(sizeof(c_char_p) // word_size - 1)
    action(DUP)
number(LIB_C_STRLEN)
action(LIB_C)

step(addr=VM.here + 1)
arg1len = S.pop()
print("arg 1's length is {}, and should be {}".format(arg1len, len(args[1])))
if arg1len != len(args[1]):
    print("Error in extra instructions tests: PC = {:#x}".format(PC.get()))
    sys.exit(1)
S.push(arg1len) # push length back for next test

# Test LIB_C_STRNCPY
number(LIB_C_STRNCPY)
action(LIB_C)

step(addr=VM.here + 1)
c_str = string_at(libsmite.smite_native_address_of_range(VM.state, buffer, 0))
print("arg 1 is {}, and should be {}".format(c_str, args[1]))
if c_str != args[1]:
    print("Error in extra instructions tests: PC = {:#x}".format(PC.get()))
    sys.exit(1)

print("Extra instructions tests ran OK")
